### PR TITLE
Implement --user option for 'kubectl exec'

### LINKED ISSUE #30656: Support the user flag from docker exec in kubectl exec
It looks like `docker exec` is being used as the backend for `kubectl exec`. `docker exec` has the `--user` flag, which allows you to run a command as a particular user.  This same functionality doesn't exist in Kubernetes.  

Our use case is that we spin up pods, and execute untrusted code in them.  However, there are times when after creating the pod, we need to run programs that need root access (they need to access privileged ports, etc).

We don't want to run the untrusted code as root in the container, which prevents us from just escalating permissions for all programs.

I looked around for references to this problem, but only found this StackOverflow answer from last year -- http://stackoverflow.com/questions/33293265/execute-command-into-kubernetes-pod-as-other-user .

There are some workarounds to this, such as setting up a server in the container that takes commands in, or defaulting to root, but dropping to another user before running untrusted code.  However, these workarounds break nice Kubernetes/Docker abstractions and introduce security holes.

### PR DESCRIPTION
**Pull request by louyihua**
**What this PR does / why we need it**:
This patch implements the `--user` option in `kubectl exec` command. With this patch, user can specify the user name/id in a `kubectl exec` command like `kubectl exec -it --user root pod-name command`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #30656

**Release note**:

```release-note
Implement --user option for 'kubectl exec'
```

### CODE REVIEW

File: pkg/kubectl/cmd/exec.go [hunk 1]
```diff
@@ -86,7 +86,7 @@ func NewCmdExec(f cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *c
 		},
 	}
 	cmd.Flags().StringVarP(&options.PodName, "pod", "p", "", "Pod name")
-	// TODO support UID
```
Comment (ncdc) [inline pkg/kubectl/cmd/exec.go]: This comment is about pod uid, not user uid

**Review by ncdc — COMMENTED** [review]
Please see the comments on #30656. That needs to be resolved first as to whether this is even something we can support, from a CRI perspective.

**Comment by ncdc** [comment]
FYI @liggitt @yujuhong @derekwaynecarr @dchen1107 @Random-Liu @kubernetes/sig-node-pr-reviews

**Comment by Nowaker** [comment]
@louyihua Would you mind finishing this up?

**Review by resouer — CHANGES_REQUESTED** [review]
This PR changes the whole path from api, kubelet to cri, but lacks proper e2e test.

**Comment by davidxia** [comment]
@louyihua were you planning on following up on this? If not, I can help take over if you'd like. If no response, I'll get started. :)

**Comment by tsunamireversal** [comment]
Can we have a status on that ? Feature would be super useful

**Comment by liggitt** [comment]
this change needs to have a associated [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/0001-kubernetes-enhancement-proposal-process.md) in implementable state before being completed/reviewed

**Comment by Nowaker** [comment]
It's a TOP3 most upvoted feature request... Can someone finish it up?

**Comment by smarterclayton** [comment]
As mentioned above this really needs a KEP and discussion about the security implications.  It’s not that it’s necessarily a bad idea, it just has a significant impact to the system and needs a design before we can start coding.

I’d be happy to help review and shepherd a KEP for this, but this PR isnt really actionable yet.

For now,

/close

And we can follow up with a KEP

**Comment by tsunamireversal** [comment]
I don't understand your comment @smarterclayton 
Adding the --user feature would be much better for security, not worse. 
Unless we're missing something...?

Also, what's a KEP ?

**Comment by SimonHeimberg** [comment]
There was a KEP: https://github.com/kubernetes/enhancements/pull/1224

**Pull request closed without merging**